src/main/java/org/folio/rest/impl/FinanceHelper.java

```diff
@@ -72,11 +85,61 @@
     List<Transaction> encumbrances = encumbranceDistributionPairs.stream()
       .map(Pair::getLeft)
       .collect(toList());
-    return setCurrentFiscalYear(encumbrances)
-      .thenCompose(aVoid -> createSummary(compPo.getId(), encumbrances.size()))
-      .thenCompose(vVoid -> createEncumbrances(encumbranceDistributionPairs));
+    return retrieveSystemCurrency()
+      .thenCompose(v -> setCurrentFiscalYear(encumbrances))
+      .thenCompose(v -> checkEnoughMoneyForTransactions(encumbrances))
+      .thenCompose(v -> createSummary(compPo.getId(), encumbrances.size()))
+      .thenCompose(summaryId -> createEncumbrances(encumbranceDistributionPairs));
   }
 
+  private CompletableFuture<Void> retrieveSystemCurrency() {
+    return getSystemCurrency().thenAccept(currency -> systemCurrency = currency);
+  }
+
+  private CompletableFuture<Void> checkEnoughMoneyForTransactions(List<Transaction> encumbrances) {
+    Map<String, List<Transaction>> groupedByFund = encumbrances.stream()
```

Comment:
Not sure why you need a map here when all you really care about is the unique set of fundIds.  I'm guessing it was a matter of convenience.  Anyway, not a huge deal.

Reply:
FWIW
```
Set<String> fundIds = new HashSet<>();
encumbrances.stream().forEach(e -> fundIds.add(e.getFromFundId()));
```
is an order of magnitude much faster, not that either impl is at all slow.